### PR TITLE
Added structure function squared functionality

### PR DESCRIPTION
**Pull request by nevencaplar**
Added a structure function functionality. This is a function commonly used to quantify variability of stochastic sources. This is following the algorithm written created by Weixiang Yu (see notebook here: https://github.com/ywx649999311/astro_code/blob/master/variability/SF.ipynb)

I have also included extensive TODO list with the functionality that would need to be included. 

Co-authored-by: Weixiang Yu <wy73@drexel.edu>

### CODE REVIEW

File: lsstseries/analysis/structurefunction2.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    flux : `numpy.ndarray` (N,)
+        Array of flux/magnitude measurements.
+    err : `numpy.ndarray` (N,)
+        Array of associated flux/magnitude errors.
+    bins : `numpy.ndarray` (N,)
+        Array of time bins in which to calculate structure function
+    band : `numpy.ndarray` (N,)
+        Array of associated band labels,
+    band_to_calc : `str` or `list` of `str`
+        Bands to calculate StetsonJ on. Single band descriptor, or list
```
Comment (dougbrn) [inline lsstseries/analysis/structurefunction2.py:21]: Adjust description to be structurefunction2-specific

File: lsstseries/analysis/structurefunction2.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+            sf2[b] = _sf2_single(times, fluxes, errors, bins)
+        else:
+            sf2[b] = np.nan
+    return sf2
+
+
+def _sf2_single(times, fluxes, errors, bins):
+    """Calculate structure function squared
+
+    Calculate structure function squared from the avaliable data. This is
```
Comment (dougbrn) [inline lsstseries/analysis/structurefunction2.py:58]: avaliable -> available

File: lsstseries/analysis/structurefunction2.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    AGN-type definition, i.e., variance of the distribution of the
+    differences of measurments at different times
```
Comment (dougbrn) [inline lsstseries/analysis/structurefunction2.py:60]: measurments -> measurements

Comment (dougbrn) [reply lsstseries/analysis/structurefunction2.py:60]: A few more occurrences of this in the docstring

File: lsstseries/analysis/structurefunction2.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        Array of associated flux/magnitude errors.
+    bins : `numpy.ndarray` (N,)
+        Array of time bins in which to calculate structure function
+    band : `numpy.ndarray` (N,)
+        Array of associated band labels,
+    band_to_calc : `str` or `list` of `str`
+        Bands to calculate StetsonJ on. Single band descriptor, or list
+        of such descriptors.
+
+    Returns
```
Comment (delucchi-cmu) [inline lsstseries/analysis/structurefunction2.py:24]: Returns array of things?

File: lsstseries/analysis/structurefunction2.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    - allow multiple inputs, with same <t> at once
+    - ability to create SF2 from multiple lightcurves at once (ensamble)
+    - allow for different definitions of SF2
+    """
+
+    times = times.values
+    fluxes = fluxes.values
+    errors = errors.values
+
+    # compute dt (difference of times) and
```
Comment (delucchi-cmu) [inline lsstseries/analysis/structurefunction2.py:93]: This comment is confusing in a couple ways:
- `difference of times` gets two abbreviations, `dt` and `d_times`
- there's no `dm` (but there is a `d_fluxes` below which I think you're talking about here)

File: lsstseries/analysis/structurefunction2.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    err2_matrix = errors.reshape((1, errors.size))**2 \
+        + errors.reshape((errors.size, 1))**2
+    err2s = err2_matrix[dt_matrix > 0].flatten().astype(np.float16)
+
+    # corrected each pair of observations
+    cor_flux2 = d_fluxes**2 - err2s
+
+    # structure function at specific dt
+    # the line below will throw error if the bins are not covering the whole range
+    SFs, bin_edgs, _ = binned_statistic(d_times, cor_flux2, 'mean', bins)
```
Comment (delucchi-cmu) [inline lsstseries/analysis/structurefunction2.py:113]: Keep variables starting with lower case.

**Review by delucchi-cmu — COMMENTED** [review]
I only have comments about the comments because I don't know any of the math...

File: lsstseries/analysis/structurefunction2.py (hunk 1 above)
Comment (nevencaplar) [reply lsstseries/analysis/structurefunction2.py:21]: Corrected

File: lsstseries/analysis/structurefunction2.py (hunk 2 above)
Comment (nevencaplar) [reply lsstseries/analysis/structurefunction2.py:58]: Fixed. I am still searching for a spell checker that I could use only on the comments....

File: lsstseries/analysis/structurefunction2.py (hunk 3 above)
Comment (nevencaplar) [reply lsstseries/analysis/structurefunction2.py:60]: fixed

File: lsstseries/analysis/structurefunction2.py (hunk 5 above)
Comment (nevencaplar) [reply lsstseries/analysis/structurefunction2.py:93]: Indeed. Fixed.

**Review by dougbrn — APPROVED** [review]
I'm also limited on reviewing the math, but this looks good with the comment tidy-ups.

**Pull request merged**